spec/support/cmap.rb

```diff
@@ -0,0 +1,346 @@
+require 'support/connection_monitoring_and_pooling/verifier'
+
+module Mongo
+  module CMAP
+
+    # Represents a specification.
+    #
+    # @since 2.7.0
+    class Spec
+
+      # @return [ String ] description The spec description.
+      attr_reader :description
+
+      # @return [ Hash ] pool_options The options for the created pools.
+      attr_reader :pool_options
+
+      # @return [ Integer ] num_pools The number of pools to create.
+      attr_reader :num_pools
+
+      # @return [ Array<Operation> ] spec_ops The spec operations.
+      attr_reader :spec_ops
+
+      # @return [ Array<Operation> ] processed_ops The processed operations.
+      attr_reader :processed_ops
+
+      # @return [ Error | nil ] error The expected error.
+      attr_reader :error
+
+      # @return [ Array<Event::Base> ] events The events expected to occur.
+      attr_reader :events
+
+      # @return [ Array<String> ] events The names of events to ignore.
+      attr_reader :ignore
+
+      # Instantiate the new spec.
+      #
+      # @example Create the spec.
+      #   Spec.new(file)
+      #
+      # @param [ String ] file The name of the file.
+      #
+      # @since 2.7.0
+      def initialize(file)
+        @test = YAML.load(ERB.new(File::read(file)).result)
+
+        @description = @test['description']
+        @pool_options = snakeize_hash(process_options(@test['poolOptions'])).tap do |opts|
+          # The CMAP spec defines the default minPoolSize as 0, but the existing driver default is
+          # 1. Until we can change it, we need to manually specify their default when the YAML
+          # doesn't specify any.
+          opts[:min_pool_size] ||= 0
+        end
+        @num_pools = @test['numberOfPools'] || 1
+        @spec_ops = @test['operations'].map { |o| Operation.new(o) }
+        @processed_ops = []
+        @error = @test['error']
+        @events = @test['events']
+        @ignore = @test['ignore'] || []
+      end
+
+      def run(cluster)
+        subscriber = EventSubscriber.new
+
+        monitoring = Mongo::Monitoring.new(monitoring: false)
+        monitoring.subscribe(Mongo::Monitoring::CONNECTION_POOL, subscriber)
+
+        server = Mongo::Server.new(
+            Address.new(SpecConfig.instance.addresses.first),
+            cluster,
+            monitoring,
+            Mongo::Event::Listeners.new,
+            pool_options)
+
+        state = {
+          'pool' => Mongo::Server::ConnectionPool.get(server) do
+            Mongo::Server::Connection.new(server)
+          end
+        }
+
+        preprocess
```

Comment:
Running a test should not mutate its expectations. If any expectations need to be altered this should happen during construction.